{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import math\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "sys.path.append('../src/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>forward</th>\n",
       "      <th>reverse</th>\n",
       "      <th>sample</th>\n",
       "      <th>sample_library</th>\n",
       "      <th>sequencing_tech</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>/valhalla/fastq/2021.03.19/21-256229-SCV-02262...</td>\n",
       "      <td>/valhalla/fastq/2021.03.19/21-256229-SCV-02262...</td>\n",
       "      <td>21-256229-SCV-022621-AM</td>\n",
       "      <td>21-256229-SCV-022621-AM_L001</td>\n",
       "      <td>illumina</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>/valhalla/fastq/2021.03.19/21-256232-SCV-02262...</td>\n",
       "      <td>/valhalla/fastq/2021.03.19/21-256232-SCV-02262...</td>\n",
       "      <td>21-256232-SCV-022621-AM</td>\n",
       "      <td>21-256232-SCV-022621-AM_L001</td>\n",
       "      <td>illumina</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>/valhalla/fastq/2021.03.19/21-256234-SCV-02262...</td>\n",
       "      <td>/valhalla/fastq/2021.03.19/21-256234-SCV-02262...</td>\n",
       "      <td>21-256234-SCV-022621-AM</td>\n",
       "      <td>21-256234-SCV-022621-AM_L001</td>\n",
       "      <td>illumina</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>/valhalla/fastq/2021.03.19/21-256237-SCV-02262...</td>\n",
       "      <td>/valhalla/fastq/2021.03.19/21-256237-SCV-02262...</td>\n",
       "      <td>21-256237-SCV-022621-AM</td>\n",
       "      <td>21-256237-SCV-022621-AM_L001</td>\n",
       "      <td>illumina</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>/valhalla/fastq/2021.03.19/21-256246-SCV-02262...</td>\n",
       "      <td>/valhalla/fastq/2021.03.19/21-256246-SCV-02262...</td>\n",
       "      <td>21-256246-SCV-022621-AM</td>\n",
       "      <td>21-256246-SCV-022621-AM_L001</td>\n",
       "      <td>illumina</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>/valhalla/fastq/2021.03.19/21-256265-SCV-02262...</td>\n",
       "      <td>/valhalla/fastq/2021.03.19/21-256265-SCV-02262...</td>\n",
       "      <td>21-256265-SCV-022621-AM</td>\n",
       "      <td>21-256265-SCV-022621-AM_L001</td>\n",
       "      <td>illumina</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>/valhalla/fastq/2021.03.19/21-256267-SCV-02262...</td>\n",
       "      <td>/valhalla/fastq/2021.03.19/21-256267-SCV-02262...</td>\n",
       "      <td>21-256267-SCV-022621-AM</td>\n",
       "      <td>21-256267-SCV-022621-AM_L001</td>\n",
       "      <td>illumina</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>/valhalla/fastq/2021.03.19/21-256274-SCV-02262...</td>\n",
       "      <td>/valhalla/fastq/2021.03.19/21-256274-SCV-02262...</td>\n",
       "      <td>21-256274-SCV-022621-AM</td>\n",
       "      <td>21-256274-SCV-022621-AM_L001</td>\n",
       "      <td>illumina</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>/valhalla/fastq/2021.03.19/Fastq/21-170487-SCV...</td>\n",
       "      <td>/valhalla/fastq/2021.03.19/Fastq/21-170487-SCV...</td>\n",
       "      <td>21-170487-SCV2-031521-AM-1-V2</td>\n",
       "      <td>21-170487-SCV2-031521-AM-1-V2_L001</td>\n",
       "      <td>illumina</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>/valhalla/fastq/2021.03.19/Fastq/21-227289-SCV...</td>\n",
       "      <td>/valhalla/fastq/2021.03.19/Fastq/21-227289-SCV...</td>\n",
       "      <td>21-227289-SCV2-031521-AM-1-V2</td>\n",
       "      <td>21-227289-SCV2-031521-AM-1-V2_L001</td>\n",
       "      <td>illumina</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>/valhalla/fastq/2021.03.19/Fastq/21-295987-SCV...</td>\n",
       "      <td>/valhalla/fastq/2021.03.19/Fastq/21-295987-SCV...</td>\n",
       "      <td>21-295987-SCV2-031521-AM-1-V2</td>\n",
       "      <td>21-295987-SCV2-031521-AM-1-V2_L001</td>\n",
       "      <td>illumina</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>/valhalla/fastq/2021.03.19/Fastq/21-295989-SCV...</td>\n",
       "      <td>/valhalla/fastq/2021.03.19/Fastq/21-295989-SCV...</td>\n",
       "      <td>21-295989-SCV2-031521-AM-1-V2</td>\n",
       "      <td>21-295989-SCV2-031521-AM-1-V2_L001</td>\n",
       "      <td>illumina</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>/valhalla/fastq/2021.03.19/Fastq/21-298438-SCV...</td>\n",
       "      <td>/valhalla/fastq/2021.03.19/Fastq/21-298438-SCV...</td>\n",
       "      <td>21-298438-SCV2-031521-AM-1-V2</td>\n",
       "      <td>21-298438-SCV2-031521-AM-1-V2_L001</td>\n",
       "      <td>illumina</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>/valhalla/fastq/2021.03.19/Fastq/21-298441-SCV...</td>\n",
       "      <td>/valhalla/fastq/2021.03.19/Fastq/21-298441-SCV...</td>\n",
       "      <td>21-298441-SCV2-031521-AM-1-V2</td>\n",
       "      <td>21-298441-SCV2-031521-AM-1-V2_L001</td>\n",
       "      <td>illumina</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>/valhalla/fastq/2021.03.19/Fastq/21-299032-SCV...</td>\n",
       "      <td>/valhalla/fastq/2021.03.19/Fastq/21-299032-SCV...</td>\n",
       "      <td>21-299032-SCV2-031521-AM-1-V2</td>\n",
       "      <td>21-299032-SCV2-031521-AM-1-V2_L001</td>\n",
       "      <td>illumina</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>/valhalla/fastq/2021.03.19/Fastq/21-299112-SCV...</td>\n",
       "      <td>/valhalla/fastq/2021.03.19/Fastq/21-299112-SCV...</td>\n",
       "      <td>21-299112-SCV2-031521-AM-1-V2</td>\n",
       "      <td>21-299112-SCV2-031521-AM-1-V2_L001</td>\n",
       "      <td>illumina</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>/valhalla/fastq/2021.03.19/Fastq/21-299114-SCV...</td>\n",
       "      <td>/valhalla/fastq/2021.03.19/Fastq/21-299114-SCV...</td>\n",
       "      <td>21-299114-SCV2-031521-AM-1-V2</td>\n",
       "      <td>21-299114-SCV2-031521-AM-1-V2_L001</td>\n",
       "      <td>illumina</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>/valhalla/fastq/2021.03.19/Fastq/21-299332-SCV...</td>\n",
       "      <td>/valhalla/fastq/2021.03.19/Fastq/21-299332-SCV...</td>\n",
       "      <td>21-299332-SCV2-031521-AM-1-V2</td>\n",
       "      <td>21-299332-SCV2-031521-AM-1-V2_L001</td>\n",
       "      <td>illumina</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>/valhalla/fastq/2021.03.19/Fastq/NTC-SCV2-0315...</td>\n",
       "      <td>/valhalla/fastq/2021.03.19/Fastq/NTC-SCV2-0315...</td>\n",
       "      <td>NTC-SCV2-031521-AM-1-V2</td>\n",
       "      <td>NTC-SCV2-031521-AM-1-V2_L001</td>\n",
       "      <td>illumina</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              forward  \\\n",
       "0   /valhalla/fastq/2021.03.19/21-256229-SCV-02262...   \n",
       "1   /valhalla/fastq/2021.03.19/21-256232-SCV-02262...   \n",
       "2   /valhalla/fastq/2021.03.19/21-256234-SCV-02262...   \n",
       "3   /valhalla/fastq/2021.03.19/21-256237-SCV-02262...   \n",
       "4   /valhalla/fastq/2021.03.19/21-256246-SCV-02262...   \n",
       "5   /valhalla/fastq/2021.03.19/21-256265-SCV-02262...   \n",
       "6   /valhalla/fastq/2021.03.19/21-256267-SCV-02262...   \n",
       "7   /valhalla/fastq/2021.03.19/21-256274-SCV-02262...   \n",
       "8   /valhalla/fastq/2021.03.19/Fastq/21-170487-SCV...   \n",
       "9   /valhalla/fastq/2021.03.19/Fastq/21-227289-SCV...   \n",
       "10  /valhalla/fastq/2021.03.19/Fastq/21-295987-SCV...   \n",
       "11  /valhalla/fastq/2021.03.19/Fastq/21-295989-SCV...   \n",
       "12  /valhalla/fastq/2021.03.19/Fastq/21-298438-SCV...   \n",
       "13  /valhalla/fastq/2021.03.19/Fastq/21-298441-SCV...   \n",
       "14  /valhalla/fastq/2021.03.19/Fastq/21-299032-SCV...   \n",
       "15  /valhalla/fastq/2021.03.19/Fastq/21-299112-SCV...   \n",
       "16  /valhalla/fastq/2021.03.19/Fastq/21-299114-SCV...   \n",
       "17  /valhalla/fastq/2021.03.19/Fastq/21-299332-SCV...   \n",
       "18  /valhalla/fastq/2021.03.19/Fastq/NTC-SCV2-0315...   \n",
       "\n",
       "                                              reverse  \\\n",
       "0   /valhalla/fastq/2021.03.19/21-256229-SCV-02262...   \n",
       "1   /valhalla/fastq/2021.03.19/21-256232-SCV-02262...   \n",
       "2   /valhalla/fastq/2021.03.19/21-256234-SCV-02262...   \n",
       "3   /valhalla/fastq/2021.03.19/21-256237-SCV-02262...   \n",
       "4   /valhalla/fastq/2021.03.19/21-256246-SCV-02262...   \n",
       "5   /valhalla/fastq/2021.03.19/21-256265-SCV-02262...   \n",
       "6   /valhalla/fastq/2021.03.19/21-256267-SCV-02262...   \n",
       "7   /valhalla/fastq/2021.03.19/21-256274-SCV-02262...   \n",
       "8   /valhalla/fastq/2021.03.19/Fastq/21-170487-SCV...   \n",
       "9   /valhalla/fastq/2021.03.19/Fastq/21-227289-SCV...   \n",
       "10  /valhalla/fastq/2021.03.19/Fastq/21-295987-SCV...   \n",
       "11  /valhalla/fastq/2021.03.19/Fastq/21-295989-SCV...   \n",
       "12  /valhalla/fastq/2021.03.19/Fastq/21-298438-SCV...   \n",
       "13  /valhalla/fastq/2021.03.19/Fastq/21-298441-SCV...   \n",
       "14  /valhalla/fastq/2021.03.19/Fastq/21-299032-SCV...   \n",
       "15  /valhalla/fastq/2021.03.19/Fastq/21-299112-SCV...   \n",
       "16  /valhalla/fastq/2021.03.19/Fastq/21-299114-SCV...   \n",
       "17  /valhalla/fastq/2021.03.19/Fastq/21-299332-SCV...   \n",
       "18  /valhalla/fastq/2021.03.19/Fastq/NTC-SCV2-0315...   \n",
       "\n",
       "                           sample                      sample_library  \\\n",
       "0         21-256229-SCV-022621-AM        21-256229-SCV-022621-AM_L001   \n",
       "1         21-256232-SCV-022621-AM        21-256232-SCV-022621-AM_L001   \n",
       "2         21-256234-SCV-022621-AM        21-256234-SCV-022621-AM_L001   \n",
       "3         21-256237-SCV-022621-AM        21-256237-SCV-022621-AM_L001   \n",
       "4         21-256246-SCV-022621-AM        21-256246-SCV-022621-AM_L001   \n",
       "5         21-256265-SCV-022621-AM        21-256265-SCV-022621-AM_L001   \n",
       "6         21-256267-SCV-022621-AM        21-256267-SCV-022621-AM_L001   \n",
       "7         21-256274-SCV-022621-AM        21-256274-SCV-022621-AM_L001   \n",
       "8   21-170487-SCV2-031521-AM-1-V2  21-170487-SCV2-031521-AM-1-V2_L001   \n",
       "9   21-227289-SCV2-031521-AM-1-V2  21-227289-SCV2-031521-AM-1-V2_L001   \n",
       "10  21-295987-SCV2-031521-AM-1-V2  21-295987-SCV2-031521-AM-1-V2_L001   \n",
       "11  21-295989-SCV2-031521-AM-1-V2  21-295989-SCV2-031521-AM-1-V2_L001   \n",
       "12  21-298438-SCV2-031521-AM-1-V2  21-298438-SCV2-031521-AM-1-V2_L001   \n",
       "13  21-298441-SCV2-031521-AM-1-V2  21-298441-SCV2-031521-AM-1-V2_L001   \n",
       "14  21-299032-SCV2-031521-AM-1-V2  21-299032-SCV2-031521-AM-1-V2_L001   \n",
       "15  21-299112-SCV2-031521-AM-1-V2  21-299112-SCV2-031521-AM-1-V2_L001   \n",
       "16  21-299114-SCV2-031521-AM-1-V2  21-299114-SCV2-031521-AM-1-V2_L001   \n",
       "17  21-299332-SCV2-031521-AM-1-V2  21-299332-SCV2-031521-AM-1-V2_L001   \n",
       "18        NTC-SCV2-031521-AM-1-V2        NTC-SCV2-031521-AM-1-V2_L001   \n",
       "\n",
       "   sequencing_tech  \n",
       "0         illumina  \n",
       "1         illumina  \n",
       "2         illumina  \n",
       "3         illumina  \n",
       "4         illumina  \n",
       "5         illumina  \n",
       "6         illumina  \n",
       "7         illumina  \n",
       "8         illumina  \n",
       "9         illumina  \n",
       "10        illumina  \n",
       "11        illumina  \n",
       "12        illumina  \n",
       "13        illumina  \n",
       "14        illumina  \n",
       "15        illumina  \n",
       "16        illumina  \n",
       "17        illumina  \n",
       "18        illumina  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "s_fp = '/valhalla/fastq/2021.03.19/sd_county_samples.tsv'\n",
    "s_df = pd.read_csv(s_fp, sep='\\t')\n",
    "s_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "s_df = s_df.loc[s_df['sample']!='21-299332-SCV2-031521-AM-1-V2']\n",
    "s_df.to_csv(s_fp, sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/al/anaconda3/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3146: DtypeWarning: Columns (7,51) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n"
     ]
    }
   ],
   "source": [
    "muts_fp = '/valhalla/gisaid/workflow/mutations_2021-03-19.csv'\n",
    "muts = pd.read_csv(muts_fp)\n",
    "# muts.loc[muts['gene'].str.contains('Non')]['gene']\n",
    "# muts.loc[~(muts['gene']=='Non-coding region'), 'gene'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>idx</th>\n",
       "      <th>seq_len</th>\n",
       "      <th>replacements</th>\n",
       "      <th>pos</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_start_pos</th>\n",
       "      <th>codon_num</th>\n",
       "      <th>codon_start</th>\n",
       "      <th>ref_codon</th>\n",
       "      <th>alt_codon</th>\n",
       "      <th>...</th>\n",
       "      <th>division_lower</th>\n",
       "      <th>location_original_lower</th>\n",
       "      <th>location_lower</th>\n",
       "      <th>tmp</th>\n",
       "      <th>date_modified</th>\n",
       "      <th>country_id</th>\n",
       "      <th>tmp_info1</th>\n",
       "      <th>division_id</th>\n",
       "      <th>tmp_info2</th>\n",
       "      <th>location_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Ireland/G-NVRL-e84IRL94605/2021</td>\n",
       "      <td>29674</td>\n",
       "      <td>822:T</td>\n",
       "      <td>822</td>\n",
       "      <td>ORF1a</td>\n",
       "      <td>265</td>\n",
       "      <td>186</td>\n",
       "      <td>820.0</td>\n",
       "      <td>GTC</td>\n",
       "      <td>GTT</td>\n",
       "      <td>...</td>\n",
       "      <td>galway</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>['2021', '02', '01']</td>\n",
       "      <td>2021-03-19-09:00</td>\n",
       "      <td>IRL</td>\n",
       "      <td>Ireland-Galway</td>\n",
       "      <td>GY</td>\n",
       "      <td>Ireland-Galway-None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Ireland/G-NVRL-e84IRL94605/2021</td>\n",
       "      <td>29674</td>\n",
       "      <td>912:T</td>\n",
       "      <td>912</td>\n",
       "      <td>ORF1a</td>\n",
       "      <td>265</td>\n",
       "      <td>216</td>\n",
       "      <td>910.0</td>\n",
       "      <td>TCC</td>\n",
       "      <td>TCT</td>\n",
       "      <td>...</td>\n",
       "      <td>galway</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>['2021', '02', '01']</td>\n",
       "      <td>2021-03-19-09:00</td>\n",
       "      <td>IRL</td>\n",
       "      <td>Ireland-Galway</td>\n",
       "      <td>GY</td>\n",
       "      <td>Ireland-Galway-None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Ireland/G-NVRL-e84IRL94605/2021</td>\n",
       "      <td>29674</td>\n",
       "      <td>3036:T</td>\n",
       "      <td>3036</td>\n",
       "      <td>ORF1a</td>\n",
       "      <td>265</td>\n",
       "      <td>924</td>\n",
       "      <td>3034.0</td>\n",
       "      <td>TTC</td>\n",
       "      <td>TTT</td>\n",
       "      <td>...</td>\n",
       "      <td>galway</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>['2021', '02', '01']</td>\n",
       "      <td>2021-03-19-09:00</td>\n",
       "      <td>IRL</td>\n",
       "      <td>Ireland-Galway</td>\n",
       "      <td>GY</td>\n",
       "      <td>Ireland-Galway-None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Ireland/G-NVRL-e84IRL94605/2021</td>\n",
       "      <td>29674</td>\n",
       "      <td>3266:T</td>\n",
       "      <td>3266</td>\n",
       "      <td>ORF1a</td>\n",
       "      <td>265</td>\n",
       "      <td>1001</td>\n",
       "      <td>3265.0</td>\n",
       "      <td>ACT</td>\n",
       "      <td>ATT</td>\n",
       "      <td>...</td>\n",
       "      <td>galway</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>['2021', '02', '01']</td>\n",
       "      <td>2021-03-19-09:00</td>\n",
       "      <td>IRL</td>\n",
       "      <td>Ireland-Galway</td>\n",
       "      <td>GY</td>\n",
       "      <td>Ireland-Galway-None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Ireland/G-NVRL-e84IRL94605/2021</td>\n",
       "      <td>29674</td>\n",
       "      <td>5387:A</td>\n",
       "      <td>5387</td>\n",
       "      <td>ORF1a</td>\n",
       "      <td>265</td>\n",
       "      <td>1708</td>\n",
       "      <td>5386.0</td>\n",
       "      <td>GCT</td>\n",
       "      <td>GAT</td>\n",
       "      <td>...</td>\n",
       "      <td>galway</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>['2021', '02', '01']</td>\n",
       "      <td>2021-03-19-09:00</td>\n",
       "      <td>IRL</td>\n",
       "      <td>Ireland-Galway</td>\n",
       "      <td>GY</td>\n",
       "      <td>Ireland-Galway-None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15150228</th>\n",
       "      <td>Belgium/ULG-10352/2020</td>\n",
       "      <td>29409</td>\n",
       "      <td>None</td>\n",
       "      <td>266</td>\n",
       "      <td>ORF1a</td>\n",
       "      <td>267</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>GGAGAGCCTTGTCCCTGGTTTCAACGAGAAAACACACGTCCAACTC...</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>liège</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>['2020', '08', '04']</td>\n",
       "      <td>2021-03-19-09:00</td>\n",
       "      <td>BEL</td>\n",
       "      <td>Belgium-Liège</td>\n",
       "      <td>None</td>\n",
       "      <td>Belgium-Liège-None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15150229</th>\n",
       "      <td>env/Uruguay/MO-1592/2020</td>\n",
       "      <td>29409</td>\n",
       "      <td>None</td>\n",
       "      <td>29666</td>\n",
       "      <td>ORF10</td>\n",
       "      <td>29559</td>\n",
       "      <td>37</td>\n",
       "      <td>None</td>\n",
       "      <td>CACATAGCA</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>montevideo</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>['2020', '05', '08']</td>\n",
       "      <td>2021-03-19-09:00</td>\n",
       "      <td>URY</td>\n",
       "      <td>Uruguay-Montevideo</td>\n",
       "      <td>MO</td>\n",
       "      <td>Uruguay-Montevideo-None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15150230</th>\n",
       "      <td>USA/FL-BPHL-1244/2020</td>\n",
       "      <td>29409</td>\n",
       "      <td>None</td>\n",
       "      <td>29668</td>\n",
       "      <td>ORF10</td>\n",
       "      <td>29559</td>\n",
       "      <td>38</td>\n",
       "      <td>None</td>\n",
       "      <td>CATAGCA</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>florida</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>['2020', '06', '02']</td>\n",
       "      <td>2021-03-19-09:00</td>\n",
       "      <td>USA</td>\n",
       "      <td>United States-Florida</td>\n",
       "      <td>FL</td>\n",
       "      <td>United States-Florida-None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15150231</th>\n",
       "      <td>India/MH-NIV-114444/2020</td>\n",
       "      <td>29409</td>\n",
       "      <td>None</td>\n",
       "      <td>25229</td>\n",
       "      <td>S</td>\n",
       "      <td>21564</td>\n",
       "      <td>1223</td>\n",
       "      <td>None</td>\n",
       "      <td>CTTGATTGCCATAGTA</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>maharashtra</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>['2020', '07', '14']</td>\n",
       "      <td>2021-03-19-09:00</td>\n",
       "      <td>IND</td>\n",
       "      <td>India-Maharashtra</td>\n",
       "      <td>MH</td>\n",
       "      <td>India-Maharashtra-None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15150232</th>\n",
       "      <td>USA/NM-UNM-00511/2020</td>\n",
       "      <td>29409</td>\n",
       "      <td>None</td>\n",
       "      <td>29666</td>\n",
       "      <td>ORF10</td>\n",
       "      <td>29559</td>\n",
       "      <td>37</td>\n",
       "      <td>None</td>\n",
       "      <td>CACATAGCA</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>new mexico</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>['2020', '04', '04']</td>\n",
       "      <td>2021-03-19-09:00</td>\n",
       "      <td>USA</td>\n",
       "      <td>United States-New Mexico</td>\n",
       "      <td>NM</td>\n",
       "      <td>United States-New Mexico-None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15149729 rows × 52 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      idx  seq_len replacements    pos   gene  \\\n",
       "0         Ireland/G-NVRL-e84IRL94605/2021    29674        822:T    822  ORF1a   \n",
       "1         Ireland/G-NVRL-e84IRL94605/2021    29674        912:T    912  ORF1a   \n",
       "2         Ireland/G-NVRL-e84IRL94605/2021    29674       3036:T   3036  ORF1a   \n",
       "3         Ireland/G-NVRL-e84IRL94605/2021    29674       3266:T   3266  ORF1a   \n",
       "4         Ireland/G-NVRL-e84IRL94605/2021    29674       5387:A   5387  ORF1a   \n",
       "...                                   ...      ...          ...    ...    ...   \n",
       "15150228           Belgium/ULG-10352/2020    29409         None    266  ORF1a   \n",
       "15150229         env/Uruguay/MO-1592/2020    29409         None  29666  ORF10   \n",
       "15150230            USA/FL-BPHL-1244/2020    29409         None  29668  ORF10   \n",
       "15150231         India/MH-NIV-114444/2020    29409         None  25229      S   \n",
       "15150232            USA/NM-UNM-00511/2020    29409         None  29666  ORF10   \n",
       "\n",
       "          gene_start_pos  codon_num codon_start  \\\n",
       "0                    265        186       820.0   \n",
       "1                    265        216       910.0   \n",
       "2                    265        924      3034.0   \n",
       "3                    265       1001      3265.0   \n",
       "4                    265       1708      5386.0   \n",
       "...                  ...        ...         ...   \n",
       "15150228             267          1        None   \n",
       "15150229           29559         37        None   \n",
       "15150230           29559         38        None   \n",
       "15150231           21564       1223        None   \n",
       "15150232           29559         37        None   \n",
       "\n",
       "                                                  ref_codon alt_codon  ...  \\\n",
       "0                                                       GTC       GTT  ...   \n",
       "1                                                       TCC       TCT  ...   \n",
       "2                                                       TTC       TTT  ...   \n",
       "3                                                       ACT       ATT  ...   \n",
       "4                                                       GCT       GAT  ...   \n",
       "...                                                     ...       ...  ...   \n",
       "15150228  GGAGAGCCTTGTCCCTGGTTTCAACGAGAAAACACACGTCCAACTC...      None  ...   \n",
       "15150229                                          CACATAGCA      None  ...   \n",
       "15150230                                            CATAGCA      None  ...   \n",
       "15150231                                   CTTGATTGCCATAGTA      None  ...   \n",
       "15150232                                          CACATAGCA      None  ...   \n",
       "\n",
       "         division_lower location_original_lower location_lower  \\\n",
       "0                galway                    none           none   \n",
       "1                galway                    none           none   \n",
       "2                galway                    none           none   \n",
       "3                galway                    none           none   \n",
       "4                galway                    none           none   \n",
       "...                 ...                     ...            ...   \n",
       "15150228          liège                    none           none   \n",
       "15150229     montevideo                    none           none   \n",
       "15150230        florida                    none           none   \n",
       "15150231    maharashtra                    none           none   \n",
       "15150232     new mexico                    none           none   \n",
       "\n",
       "                           tmp     date_modified country_id  \\\n",
       "0         ['2021', '02', '01']  2021-03-19-09:00        IRL   \n",
       "1         ['2021', '02', '01']  2021-03-19-09:00        IRL   \n",
       "2         ['2021', '02', '01']  2021-03-19-09:00        IRL   \n",
       "3         ['2021', '02', '01']  2021-03-19-09:00        IRL   \n",
       "4         ['2021', '02', '01']  2021-03-19-09:00        IRL   \n",
       "...                        ...               ...        ...   \n",
       "15150228  ['2020', '08', '04']  2021-03-19-09:00        BEL   \n",
       "15150229  ['2020', '05', '08']  2021-03-19-09:00        URY   \n",
       "15150230  ['2020', '06', '02']  2021-03-19-09:00        USA   \n",
       "15150231  ['2020', '07', '14']  2021-03-19-09:00        IND   \n",
       "15150232  ['2020', '04', '04']  2021-03-19-09:00        USA   \n",
       "\n",
       "                         tmp_info1 division_id                      tmp_info2  \\\n",
       "0                   Ireland-Galway          GY            Ireland-Galway-None   \n",
       "1                   Ireland-Galway          GY            Ireland-Galway-None   \n",
       "2                   Ireland-Galway          GY            Ireland-Galway-None   \n",
       "3                   Ireland-Galway          GY            Ireland-Galway-None   \n",
       "4                   Ireland-Galway          GY            Ireland-Galway-None   \n",
       "...                            ...         ...                            ...   \n",
       "15150228             Belgium-Liège        None             Belgium-Liège-None   \n",
       "15150229        Uruguay-Montevideo          MO        Uruguay-Montevideo-None   \n",
       "15150230     United States-Florida          FL     United States-Florida-None   \n",
       "15150231         India-Maharashtra          MH         India-Maharashtra-None   \n",
       "15150232  United States-New Mexico          NM  United States-New Mexico-None   \n",
       "\n",
       "         location_id  \n",
       "0               None  \n",
       "1               None  \n",
       "2               None  \n",
       "3               None  \n",
       "4               None  \n",
       "...              ...  \n",
       "15150228        None  \n",
       "15150229        None  \n",
       "15150230        None  \n",
       "15150231        None  \n",
       "15150232        None  \n",
       "\n",
       "[15149729 rows x 52 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "muts.loc[muts['date_collected']<=muts['date_submitted']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(29, 3)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(muts.loc[muts['date_collected']>muts['date_submitted']][['strain', 'date_collected', 'date_submitted']]\n",
    " .drop_duplicates(subset=['strain'])).shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "j_fp = '/home/al/analysis/gisaid/new_api_data.json.gz'\n",
    "j_df = pd.read_json(j_fp, compression='gzip')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(754926, 24)\n",
      "(754926,)\n"
     ]
    }
   ],
   "source": [
    "print(j_df.shape)\n",
    "print(j_df['accession_id'].unique().shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import bjorn_support as bs\n",
    "import mutations as bm\n",
    "# import unsupervised_learning as bul"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "m_fp = '/valhalla/gisaid/workflow/metadata_2021-03-14.tsv.gz'\n",
    "df = bs.create_chunk_names(m_fp, 10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chunk_names</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chunk_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chunk_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chunk_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chunk_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chunk_5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>chunk_74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>chunk_75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>chunk_76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>chunk_77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>chunk_78</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>78 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   chunk_names\n",
       "0      chunk_1\n",
       "1      chunk_2\n",
       "2      chunk_3\n",
       "3      chunk_4\n",
       "4      chunk_5\n",
       "..         ...\n",
       "73    chunk_74\n",
       "74    chunk_75\n",
       "75    chunk_76\n",
       "76    chunk_77\n",
       "77    chunk_78\n",
       "\n",
       "[78 rows x 1 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_chunk_names(meta_filepath: str, chunk_size: int) -> pd.DataFrame:\n",
    "    # with open(fasta_filepath, 'r') as filehandle:\n",
    "    #     count = 0\n",
    "    #     for line in filehandle:\n",
    "    #         if line.startswith(\">\"):\n",
    "    #             count += 1\n",
    "    meta_df = pd.read_csv(meta_filepath, sep='\\t', compression='gzip')\n",
    "    num_sequences = meta_df['strain'].unique().shape[0]\n",
    "    chunk_names = [f\"chunk_{i+1}\" for i in range(math.ceil(num_sequences / chunk_size))]\n",
    "    return pd.DataFrame(data=chunk_names, columns=['chunk_names'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Path('chunks/fasta/date')"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from path import Path\n",
    "t = \"chunks/fasta/date/sample.fasta\"\n",
    "Path(t).parent"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# meta_df = pd.read_csv(m_fp, sep='\\t', compression='gzip')\n",
    "# num_sequences = meta_df['accession_id'].unique().shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "770064"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "num_sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "769793"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_df['strain'].unique().shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "769794"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "760000 + 9794"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "chunk_names = [f\"chunk_{i+1}\" for i in range(math.ceil(num_sequences / 10000))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['chunk_1',\n",
       " 'chunk_2',\n",
       " 'chunk_3',\n",
       " 'chunk_4',\n",
       " 'chunk_5',\n",
       " 'chunk_6',\n",
       " 'chunk_7',\n",
       " 'chunk_8',\n",
       " 'chunk_9',\n",
       " 'chunk_10',\n",
       " 'chunk_11',\n",
       " 'chunk_12',\n",
       " 'chunk_13',\n",
       " 'chunk_14',\n",
       " 'chunk_15',\n",
       " 'chunk_16',\n",
       " 'chunk_17',\n",
       " 'chunk_18',\n",
       " 'chunk_19',\n",
       " 'chunk_20',\n",
       " 'chunk_21',\n",
       " 'chunk_22',\n",
       " 'chunk_23',\n",
       " 'chunk_24',\n",
       " 'chunk_25',\n",
       " 'chunk_26',\n",
       " 'chunk_27',\n",
       " 'chunk_28',\n",
       " 'chunk_29',\n",
       " 'chunk_30',\n",
       " 'chunk_31',\n",
       " 'chunk_32',\n",
       " 'chunk_33',\n",
       " 'chunk_34',\n",
       " 'chunk_35',\n",
       " 'chunk_36',\n",
       " 'chunk_37',\n",
       " 'chunk_38',\n",
       " 'chunk_39',\n",
       " 'chunk_40',\n",
       " 'chunk_41',\n",
       " 'chunk_42',\n",
       " 'chunk_43',\n",
       " 'chunk_44',\n",
       " 'chunk_45',\n",
       " 'chunk_46',\n",
       " 'chunk_47',\n",
       " 'chunk_48',\n",
       " 'chunk_49',\n",
       " 'chunk_50',\n",
       " 'chunk_51',\n",
       " 'chunk_52',\n",
       " 'chunk_53',\n",
       " 'chunk_54',\n",
       " 'chunk_55',\n",
       " 'chunk_56',\n",
       " 'chunk_57',\n",
       " 'chunk_58',\n",
       " 'chunk_59',\n",
       " 'chunk_60',\n",
       " 'chunk_61',\n",
       " 'chunk_62',\n",
       " 'chunk_63',\n",
       " 'chunk_64',\n",
       " 'chunk_65',\n",
       " 'chunk_66',\n",
       " 'chunk_67',\n",
       " 'chunk_68',\n",
       " 'chunk_69',\n",
       " 'chunk_70',\n",
       " 'chunk_71',\n",
       " 'chunk_72',\n",
       " 'chunk_73',\n",
       " 'chunk_74',\n",
       " 'chunk_75',\n",
       " 'chunk_76',\n",
       " 'chunk_77',\n",
       " 'chunk_78']"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chunk_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "m_fp = '/valhalla/gisaid/workflow/metadata_2021-03-13.tsv.gz'\n",
    "meta = pd.read_csv(m_fp, sep='\\t', compression='gzip')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strain</th>\n",
       "      <th>covsurver_prot_mutations</th>\n",
       "      <th>location</th>\n",
       "      <th>pangolin_lineage</th>\n",
       "      <th>date_collected</th>\n",
       "      <th>accession_id</th>\n",
       "      <th>pangolin_version</th>\n",
       "      <th>clade</th>\n",
       "      <th>date_submitted</th>\n",
       "      <th>country</th>\n",
       "      <th>division</th>\n",
       "      <th>country_normed</th>\n",
       "      <th>division_normed</th>\n",
       "      <th>location_normed</th>\n",
       "      <th>country_lower</th>\n",
       "      <th>country_normed_lower</th>\n",
       "      <th>division_lower</th>\n",
       "      <th>division_normed_lower</th>\n",
       "      <th>location_lower</th>\n",
       "      <th>location_normed_lower</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Australia/NT12/2020</td>\n",
       "      <td>(NSP15_A283V,NSP12_P323L,Spike_D614G)</td>\n",
       "      <td>None</td>\n",
       "      <td>B.1</td>\n",
       "      <td>2020</td>\n",
       "      <td>EPI_ISL_426900</td>\n",
       "      <td>2021-02-21</td>\n",
       "      <td>G</td>\n",
       "      <td>2020-04-17</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Northern territory</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Northern Territory</td>\n",
       "      <td>None</td>\n",
       "      <td>australia</td>\n",
       "      <td>australia</td>\n",
       "      <td>northern territory</td>\n",
       "      <td>northern territory</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Australia/NT13/2020</td>\n",
       "      <td>(NSP15_A283V,NSP12_P323L,Spike_D614G)</td>\n",
       "      <td>None</td>\n",
       "      <td>B.1</td>\n",
       "      <td>2020</td>\n",
       "      <td>EPI_ISL_426901</td>\n",
       "      <td>2021-02-21</td>\n",
       "      <td>G</td>\n",
       "      <td>2020-04-17</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Northern territory</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Northern Territory</td>\n",
       "      <td>None</td>\n",
       "      <td>australia</td>\n",
       "      <td>australia</td>\n",
       "      <td>northern territory</td>\n",
       "      <td>northern territory</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Australia/NT14/2020</td>\n",
       "      <td>(NSP14_R163C,NSP3_K38R,NS3_G251V,NSP2_I559V,NS...</td>\n",
       "      <td>None</td>\n",
       "      <td>B.40</td>\n",
       "      <td>2020</td>\n",
       "      <td>EPI_ISL_426902</td>\n",
       "      <td>2021-02-21</td>\n",
       "      <td>V</td>\n",
       "      <td>2020-04-17</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Northern territory</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Northern Territory</td>\n",
       "      <td>None</td>\n",
       "      <td>australia</td>\n",
       "      <td>australia</td>\n",
       "      <td>northern territory</td>\n",
       "      <td>northern territory</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Australia/NT16/2020</td>\n",
       "      <td>(NSP12_P323L,Spike_D614G)</td>\n",
       "      <td>None</td>\n",
       "      <td>B.1.8</td>\n",
       "      <td>2020</td>\n",
       "      <td>EPI_ISL_426903</td>\n",
       "      <td>2021-02-21</td>\n",
       "      <td>G</td>\n",
       "      <td>2020-04-17</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Northern territory</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Northern Territory</td>\n",
       "      <td>None</td>\n",
       "      <td>australia</td>\n",
       "      <td>australia</td>\n",
       "      <td>northern territory</td>\n",
       "      <td>northern territory</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Australia/NT17/2020</td>\n",
       "      <td>(NSP12_P323L,Spike_D614G)</td>\n",
       "      <td>None</td>\n",
       "      <td>B.1.8</td>\n",
       "      <td>2020</td>\n",
       "      <td>EPI_ISL_426904</td>\n",
       "      <td>2021-02-21</td>\n",
       "      <td>G</td>\n",
       "      <td>2020-04-17</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Northern territory</td>\n",
       "      <td>Australia</td>\n",
       "      <td>Northern Territory</td>\n",
       "      <td>None</td>\n",
       "      <td>australia</td>\n",
       "      <td>australia</td>\n",
       "      <td>northern territory</td>\n",
       "      <td>northern territory</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>764756</th>\n",
       "      <td>USA/ND-NDDH-0345/2021</td>\n",
       "      <td>(N_D377Y,NSP3_L72F,N_P67S,NSP16_R216C,NS3_Q57H...</td>\n",
       "      <td>None</td>\n",
       "      <td>B.1.2</td>\n",
       "      <td>2021-01-07</td>\n",
       "      <td>EPI_ISL_1233725</td>\n",
       "      <td>2021-02-21</td>\n",
       "      <td>GH</td>\n",
       "      <td>2021-03-12</td>\n",
       "      <td>USA</td>\n",
       "      <td>North Dakota</td>\n",
       "      <td>United States</td>\n",
       "      <td>North Dakota</td>\n",
       "      <td>None</td>\n",
       "      <td>usa</td>\n",
       "      <td>united states</td>\n",
       "      <td>north dakota</td>\n",
       "      <td>north dakota</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>764757</th>\n",
       "      <td>USA/ND-NDDH-0346/2021</td>\n",
       "      <td>(NSP15_A117T,NSP3_K837R,N_S194L,NSP5_P96L,NSP7...</td>\n",
       "      <td>None</td>\n",
       "      <td>B.1.400</td>\n",
       "      <td>2021-01-07</td>\n",
       "      <td>EPI_ISL_1233726</td>\n",
       "      <td>2021-02-21</td>\n",
       "      <td>G</td>\n",
       "      <td>2021-03-12</td>\n",
       "      <td>USA</td>\n",
       "      <td>North Dakota</td>\n",
       "      <td>United States</td>\n",
       "      <td>North Dakota</td>\n",
       "      <td>None</td>\n",
       "      <td>usa</td>\n",
       "      <td>united states</td>\n",
       "      <td>north dakota</td>\n",
       "      <td>north dakota</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>764758</th>\n",
       "      <td>USA/ND-NDDH-0347/2021</td>\n",
       "      <td>(N_D377Y,NSP3_K837R,NSP3_L1301F,Spike_S494P,NS...</td>\n",
       "      <td>None</td>\n",
       "      <td>B.1.2</td>\n",
       "      <td>2021-01-07</td>\n",
       "      <td>EPI_ISL_1233727</td>\n",
       "      <td>2021-02-21</td>\n",
       "      <td>GH</td>\n",
       "      <td>2021-03-12</td>\n",
       "      <td>USA</td>\n",
       "      <td>North Dakota</td>\n",
       "      <td>United States</td>\n",
       "      <td>North Dakota</td>\n",
       "      <td>None</td>\n",
       "      <td>usa</td>\n",
       "      <td>united states</td>\n",
       "      <td>north dakota</td>\n",
       "      <td>north dakota</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>764759</th>\n",
       "      <td>USA/ND-NDDH-0348/2021</td>\n",
       "      <td>(NSP3_P968L,Spike_H69del,Spike_A67V,N_A12G,E_L...</td>\n",
       "      <td>None</td>\n",
       "      <td>B.1.525</td>\n",
       "      <td>2021-02-26</td>\n",
       "      <td>EPI_ISL_1233728</td>\n",
       "      <td>2021-02-21</td>\n",
       "      <td>G</td>\n",
       "      <td>2021-03-12</td>\n",
       "      <td>USA</td>\n",
       "      <td>North Dakota</td>\n",
       "      <td>United States</td>\n",
       "      <td>North Dakota</td>\n",
       "      <td>None</td>\n",
       "      <td>usa</td>\n",
       "      <td>united states</td>\n",
       "      <td>north dakota</td>\n",
       "      <td>north dakota</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>764760</th>\n",
       "      <td>USA/ND-NDDH-0349/2021</td>\n",
       "      <td>(NSP3_P968L,Spike_H69del,Spike_A67V,N_A12G,NSP...</td>\n",
       "      <td>None</td>\n",
       "      <td>B.1.525</td>\n",
       "      <td>2021-02-25</td>\n",
       "      <td>EPI_ISL_1233729</td>\n",
       "      <td>2021-02-21</td>\n",
       "      <td>G</td>\n",
       "      <td>2021-03-12</td>\n",
       "      <td>USA</td>\n",
       "      <td>North Dakota</td>\n",
       "      <td>United States</td>\n",
       "      <td>North Dakota</td>\n",
       "      <td>None</td>\n",
       "      <td>usa</td>\n",
       "      <td>united states</td>\n",
       "      <td>north dakota</td>\n",
       "      <td>north dakota</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>764761 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       strain  \\\n",
       "0         Australia/NT12/2020   \n",
       "1         Australia/NT13/2020   \n",
       "2         Australia/NT14/2020   \n",
       "3         Australia/NT16/2020   \n",
       "4         Australia/NT17/2020   \n",
       "...                       ...   \n",
       "764756  USA/ND-NDDH-0345/2021   \n",
       "764757  USA/ND-NDDH-0346/2021   \n",
       "764758  USA/ND-NDDH-0347/2021   \n",
       "764759  USA/ND-NDDH-0348/2021   \n",
       "764760  USA/ND-NDDH-0349/2021   \n",
       "\n",
       "                                 covsurver_prot_mutations location  \\\n",
       "0                   (NSP15_A283V,NSP12_P323L,Spike_D614G)     None   \n",
       "1                   (NSP15_A283V,NSP12_P323L,Spike_D614G)     None   \n",
       "2       (NSP14_R163C,NSP3_K38R,NS3_G251V,NSP2_I559V,NS...     None   \n",
       "3                               (NSP12_P323L,Spike_D614G)     None   \n",
       "4                               (NSP12_P323L,Spike_D614G)     None   \n",
       "...                                                   ...      ...   \n",
       "764756  (N_D377Y,NSP3_L72F,N_P67S,NSP16_R216C,NS3_Q57H...     None   \n",
       "764757  (NSP15_A117T,NSP3_K837R,N_S194L,NSP5_P96L,NSP7...     None   \n",
       "764758  (N_D377Y,NSP3_K837R,NSP3_L1301F,Spike_S494P,NS...     None   \n",
       "764759  (NSP3_P968L,Spike_H69del,Spike_A67V,N_A12G,E_L...     None   \n",
       "764760  (NSP3_P968L,Spike_H69del,Spike_A67V,N_A12G,NSP...     None   \n",
       "\n",
       "       pangolin_lineage date_collected     accession_id pangolin_version  \\\n",
       "0                   B.1           2020   EPI_ISL_426900       2021-02-21   \n",
       "1                   B.1           2020   EPI_ISL_426901       2021-02-21   \n",
       "2                  B.40           2020   EPI_ISL_426902       2021-02-21   \n",
       "3                 B.1.8           2020   EPI_ISL_426903       2021-02-21   \n",
       "4                 B.1.8           2020   EPI_ISL_426904       2021-02-21   \n",
       "...                 ...            ...              ...              ...   \n",
       "764756            B.1.2     2021-01-07  EPI_ISL_1233725       2021-02-21   \n",
       "764757          B.1.400     2021-01-07  EPI_ISL_1233726       2021-02-21   \n",
       "764758            B.1.2     2021-01-07  EPI_ISL_1233727       2021-02-21   \n",
       "764759          B.1.525     2021-02-26  EPI_ISL_1233728       2021-02-21   \n",
       "764760          B.1.525     2021-02-25  EPI_ISL_1233729       2021-02-21   \n",
       "\n",
       "       clade date_submitted    country            division country_normed  \\\n",
       "0          G     2020-04-17  Australia  Northern territory      Australia   \n",
       "1          G     2020-04-17  Australia  Northern territory      Australia   \n",
       "2          V     2020-04-17  Australia  Northern territory      Australia   \n",
       "3          G     2020-04-17  Australia  Northern territory      Australia   \n",
       "4          G     2020-04-17  Australia  Northern territory      Australia   \n",
       "...      ...            ...        ...                 ...            ...   \n",
       "764756    GH     2021-03-12        USA        North Dakota  United States   \n",
       "764757     G     2021-03-12        USA        North Dakota  United States   \n",
       "764758    GH     2021-03-12        USA        North Dakota  United States   \n",
       "764759     G     2021-03-12        USA        North Dakota  United States   \n",
       "764760     G     2021-03-12        USA        North Dakota  United States   \n",
       "\n",
       "           division_normed location_normed country_lower country_normed_lower  \\\n",
       "0       Northern Territory            None     australia            australia   \n",
       "1       Northern Territory            None     australia            australia   \n",
       "2       Northern Territory            None     australia            australia   \n",
       "3       Northern Territory            None     australia            australia   \n",
       "4       Northern Territory            None     australia            australia   \n",
       "...                    ...             ...           ...                  ...   \n",
       "764756        North Dakota            None           usa        united states   \n",
       "764757        North Dakota            None           usa        united states   \n",
       "764758        North Dakota            None           usa        united states   \n",
       "764759        North Dakota            None           usa        united states   \n",
       "764760        North Dakota            None           usa        united states   \n",
       "\n",
       "            division_lower division_normed_lower location_lower  \\\n",
       "0       northern territory    northern territory           none   \n",
       "1       northern territory    northern territory           none   \n",
       "2       northern territory    northern territory           none   \n",
       "3       northern territory    northern territory           none   \n",
       "4       northern territory    northern territory           none   \n",
       "...                    ...                   ...            ...   \n",
       "764756        north dakota          north dakota           none   \n",
       "764757        north dakota          north dakota           none   \n",
       "764758        north dakota          north dakota           none   \n",
       "764759        north dakota          north dakota           none   \n",
       "764760        north dakota          north dakota           none   \n",
       "\n",
       "       location_normed_lower  \n",
       "0                       none  \n",
       "1                       none  \n",
       "2                       none  \n",
       "3                       none  \n",
       "4                       none  \n",
       "...                      ...  \n",
       "764756                  none  \n",
       "764757                  none  \n",
       "764758                  none  \n",
       "764759                  none  \n",
       "764760                  none  \n",
       "\n",
       "[764761 rows x 20 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'gs://andersen-lab_temp/outbreak_genomics/new_api_data.json.gz:\\n    Creation time:          Fri, 12 Mar 2021 18:29:33 GMT\\n    Update time:            Fri, 12 Mar 2021 18:29:38 GMT\\n    Storage class:          MULTI_REGIONAL\\n    Content-Language:       en\\n    Content-Length:         130872684\\n    Content-Type:           application/json\\n    Hash (crc32c):          YwJP2A==\\n    Hash (md5):             99G3k2Qh2pSuj/KCZtm2Iw==\\n    ETag:                   CIa3i6exq+8CEAI=\\n    Generation:             1615573773638534\\n    Metageneration:         2\\n'"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "api_data_fp = 'new_api_data.json.gz'\n",
    "gcloud_stat_cmd = f'/home/al/code/google-cloud-sdk/bin/gsutil stat gs://andersen-lab_temp/outbreak_genomics/new_api_data.json.gz'\n",
    "t = bs.run_command_log(gcloud_stat_cmd)\n",
    "t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/al/anaconda3/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3146: DtypeWarning: Columns (7,35,41,43,51) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n"
     ]
    }
   ],
   "source": [
    "fp = '/valhalla/gisaid/workflow/mutations_2021-03-09.csv'\n",
    "muts = pd.read_csv(fp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "25605        6085\n",
       "25606        6085\n",
       "25607        6085\n",
       "25608        6085\n",
       "25609        6085\n",
       "            ...  \n",
       "12838055    48225\n",
       "12838056    48225\n",
       "12838057    48225\n",
       "12838058    48225\n",
       "12838059    48225\n",
       "Name: location_id, Length: 2516606, dtype: object"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "muts.loc[muts['country']=='United States', 'location_id']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "muts = muts[muts['is_synonymous']==False]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [],
   "source": [
    "lineage = 'B.1.526'\n",
    "data = (muts.loc[muts['pangolin_lineage']==lineage, 'mutation']\n",
    "        .value_counts()\n",
    "        .to_frame()\n",
    "        .reset_index()\n",
    "        .rename(columns={'index': 'name', 'mutation': 'count'}))#.values\n",
    "data['frequency'] = (data['count'] - np.min(data['count']))/np.ptp(data['count'])\n",
    "data = data[data['frequency']>=0.05]\n",
    "# data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.spatial import distance\n",
    "from scipy.stats import mode\n",
    "from sklearn.datasets import load_breast_cancer\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.cluster import SpectralClustering\n",
    "from sklearn.metrics import confusion_matrix, accuracy_score\n",
    "from skimage.io import imread"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [],
   "source": [
    "def kmeans_unsupervised(X:np.ndarray, k:int, centroids=None, tolerance=1e-2):\n",
    "    \"\"\"K-Means Clustering with unsupervised learning trick to infer characteristic mutations\n",
    "    for a given SARS-CoV-2 lineage\"\"\"\n",
    "    # Initialize Input, Centroids, Clusters\n",
    "    if len(X.shape) < 2: X = np.expand_dims(X, axis=1)\n",
    "    if centroids: \n",
    "        centroids = bul.init_centroids(X, k)\n",
    "    else:\n",
    "        centroids = X[np.random.choice(X.shape[0], size=k, replace=False), :]\n",
    "    clusters = [[] for c in centroids]\n",
    "    # Until Convergence\n",
    "    while(True):\n",
    "        # Centroid Norm at t-1\n",
    "        old_centroids_norm = np.linalg.norm(centroids)\n",
    "        # Minimum Distance to Centroid Per Record\n",
    "        js = np.argmin(distance.cdist(X, centroids), axis=1)\n",
    "        # Assign Records to Clusters\n",
    "        for i, c in enumerate(clusters):\n",
    "             clusters[i] = list(np.where(js == i)[0])\n",
    "        # Recompute Cluster Centroids\n",
    "        for j, c in enumerate(centroids):\n",
    "            centroids[j] = np.mean(X[clusters[j]], axis=0)\n",
    "        # Centroid Norm at t\n",
    "        centroids_norm = np.linalg.norm(centroids)\n",
    "        # Convergence!\n",
    "        if np.abs(centroids_norm - old_centroids_norm) < tolerance:\n",
    "            preds = np.zeros_like(X)\n",
    "            clusters = np.asarray(clusters)[np.argsort(centroids.squeeze())[::-1]]\n",
    "#             return clusters, centroids\n",
    "            for i, c in enumerate(clusters):\n",
    "                preds[c] = i\n",
    "            return preds.squeeze()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/al/anaconda3/envs/bjorn/lib/python3.8/site-packages/numpy/core/_asarray.py:83: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray\n",
      "  return array(a, dtype, copy=False, order=order)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([ True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True, False, False, False,\n",
       "       False, False, False, False, False, False])"
      ]
     },
     "execution_count": 128,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X = np.expand_dims(data['frequency'].to_numpy(), axis=1)\n",
    "preds = kmeans_unsupervised(X, k=2, centroids=\"kmeans++\") == 0\n",
    "preds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [],
   "source": [
    "# data['is_characteristic'] = bul.spectral_clustering_unsupervised(X, num_classes=2, n_trees=100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [],
   "source": [
    "data['is_characteristic'] = preds\n",
    "data['lineage'] = lineage\n",
    "data.to_csv(f'/Users/al/Documents/scripps/analysis/bjorn/{lineage}_char_mutations.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>count</th>\n",
       "      <th>frequency</th>\n",
       "      <th>is_characteristic</th>\n",
       "      <th>lineage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>S:D614G</td>\n",
       "      <td>1621</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>True</td>\n",
       "      <td>B.1.526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>S:D253G</td>\n",
       "      <td>1620</td>\n",
       "      <td>0.999383</td>\n",
       "      <td>True</td>\n",
       "      <td>B.1.526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ORF1b:Q1011H</td>\n",
       "      <td>1618</td>\n",
       "      <td>0.998148</td>\n",
       "      <td>True</td>\n",
       "      <td>B.1.526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ORF1b:P314L</td>\n",
       "      <td>1617</td>\n",
       "      <td>0.997531</td>\n",
       "      <td>True</td>\n",
       "      <td>B.1.526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ORF1a:T265I</td>\n",
       "      <td>1611</td>\n",
       "      <td>0.993827</td>\n",
       "      <td>True</td>\n",
       "      <td>B.1.526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ORF1a:L3201P</td>\n",
       "      <td>1606</td>\n",
       "      <td>0.990741</td>\n",
       "      <td>True</td>\n",
       "      <td>B.1.526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>S:T95I</td>\n",
       "      <td>1602</td>\n",
       "      <td>0.988272</td>\n",
       "      <td>True</td>\n",
       "      <td>B.1.526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ORF3a:Q57H</td>\n",
       "      <td>1591</td>\n",
       "      <td>0.981481</td>\n",
       "      <td>True</td>\n",
       "      <td>B.1.526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ORF1a:DEL3675/3677</td>\n",
       "      <td>1574</td>\n",
       "      <td>0.970988</td>\n",
       "      <td>True</td>\n",
       "      <td>B.1.526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>S:L5F</td>\n",
       "      <td>1561</td>\n",
       "      <td>0.962963</td>\n",
       "      <td>True</td>\n",
       "      <td>B.1.526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ORF8:T11I</td>\n",
       "      <td>1534</td>\n",
       "      <td>0.946296</td>\n",
       "      <td>True</td>\n",
       "      <td>B.1.526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ORF3a:P42L</td>\n",
       "      <td>1529</td>\n",
       "      <td>0.943210</td>\n",
       "      <td>True</td>\n",
       "      <td>B.1.526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>N:M234I</td>\n",
       "      <td>1202</td>\n",
       "      <td>0.741358</td>\n",
       "      <td>True</td>\n",
       "      <td>B.1.526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>N:P199L</td>\n",
       "      <td>1199</td>\n",
       "      <td>0.739506</td>\n",
       "      <td>True</td>\n",
       "      <td>B.1.526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>S:A701V</td>\n",
       "      <td>1156</td>\n",
       "      <td>0.712963</td>\n",
       "      <td>True</td>\n",
       "      <td>B.1.526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>S:E484K</td>\n",
       "      <td>905</td>\n",
       "      <td>0.558025</td>\n",
       "      <td>False</td>\n",
       "      <td>B.1.526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ORF1a:G1946S</td>\n",
       "      <td>576</td>\n",
       "      <td>0.354938</td>\n",
       "      <td>False</td>\n",
       "      <td>B.1.526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>S:S477N</td>\n",
       "      <td>451</td>\n",
       "      <td>0.277778</td>\n",
       "      <td>False</td>\n",
       "      <td>B.1.526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>N:S202R</td>\n",
       "      <td>386</td>\n",
       "      <td>0.237654</td>\n",
       "      <td>False</td>\n",
       "      <td>B.1.526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ORF7a:L116F</td>\n",
       "      <td>380</td>\n",
       "      <td>0.233951</td>\n",
       "      <td>False</td>\n",
       "      <td>B.1.526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>N:P13L</td>\n",
       "      <td>378</td>\n",
       "      <td>0.232716</td>\n",
       "      <td>False</td>\n",
       "      <td>B.1.526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>S:Q957R</td>\n",
       "      <td>367</td>\n",
       "      <td>0.225926</td>\n",
       "      <td>False</td>\n",
       "      <td>B.1.526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>ORF1b:P2531S</td>\n",
       "      <td>215</td>\n",
       "      <td>0.132099</td>\n",
       "      <td>False</td>\n",
       "      <td>B.1.526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ORF1a:S2981F</td>\n",
       "      <td>96</td>\n",
       "      <td>0.058642</td>\n",
       "      <td>False</td>\n",
       "      <td>B.1.526</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  name  count  frequency  is_characteristic  lineage\n",
       "0              S:D614G   1621   1.000000               True  B.1.526\n",
       "1              S:D253G   1620   0.999383               True  B.1.526\n",
       "2         ORF1b:Q1011H   1618   0.998148               True  B.1.526\n",
       "3          ORF1b:P314L   1617   0.997531               True  B.1.526\n",
       "4          ORF1a:T265I   1611   0.993827               True  B.1.526\n",
       "5         ORF1a:L3201P   1606   0.990741               True  B.1.526\n",
       "6               S:T95I   1602   0.988272               True  B.1.526\n",
       "7           ORF3a:Q57H   1591   0.981481               True  B.1.526\n",
       "8   ORF1a:DEL3675/3677   1574   0.970988               True  B.1.526\n",
       "9                S:L5F   1561   0.962963               True  B.1.526\n",
       "10           ORF8:T11I   1534   0.946296               True  B.1.526\n",
       "11          ORF3a:P42L   1529   0.943210               True  B.1.526\n",
       "12             N:M234I   1202   0.741358               True  B.1.526\n",
       "13             N:P199L   1199   0.739506               True  B.1.526\n",
       "14             S:A701V   1156   0.712963               True  B.1.526\n",
       "15             S:E484K    905   0.558025              False  B.1.526\n",
       "16        ORF1a:G1946S    576   0.354938              False  B.1.526\n",
       "17             S:S477N    451   0.277778              False  B.1.526\n",
       "18             N:S202R    386   0.237654              False  B.1.526\n",
       "19         ORF7a:L116F    380   0.233951              False  B.1.526\n",
       "20              N:P13L    378   0.232716              False  B.1.526\n",
       "21             S:Q957R    367   0.225926              False  B.1.526\n",
       "22        ORF1b:P2531S    215   0.132099              False  B.1.526\n",
       "23        ORF1a:S2981F     96   0.058642              False  B.1.526"
      ]
     },
     "execution_count": 131,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bjorn",
   "language": "python",
   "name": "bjorn"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
