<!DOCTYPE html>
<html>
  <head>
    <script src="https://cdn.plot.ly/plotly-latest.min.js"></script>
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.3.1/jquery.min.js"></script>
    <link href="https://cdn.jsdelivr.net/npm/bootstrap@5.0.0-beta1/dist/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-giJF6kkoqNQ00vy+HMDP7azOuL0xtbfIcaT9wjKHr8RbDVddVHyTfAAsrekwKmP1" crossorigin="anonymous">
    <script>
      $(document).ready(function(){
      // Add smooth scrolling to all links
      $("a").on('click', function(event) {
      // Make sure this.hash has a value before overriding default behavior
      if (this.hash !== "") {
      // Prevent default anchor click behavior
      event.preventDefault();
      // Store hash
      var hash = this.hash;
      // Using jQuery's animate() method to add smooth page scroll
      // The optional number (800) specifies the number of milliseconds it takes to scroll to the specified area
      $('html, body').animate({
      scrollTop: $(hash).offset().top
      }, 800, function(){
      // Add hash (#) to URL when done scrolling (default click behavior)
      window.location.hash = hash;
      });
      } // End if
      });
      });
    </script>
  </head>
  <body>
    <nav class="navbar navbar-dark bg-dark">
      <div class="container-fluid">
	<span class="navbar-brand mb-0 h1">Situation Report (SITREP) regarding B.1.1.7 lineage in USA:</span>
      </div>
    </nav>
    <br>
    <div class="container">
      <!-- *** Section 1 - General Info *** --->
      <div class="row" id="section1">
    <hr>
    <div class="alert alert-light" role="alert">
        Concerns surrounding a new strain of SARS-CoV-2, 
      the virus behind the COVID-19 pandemic, have been developing.
      B.1.1.7, also known as Variant of Concern 202012/01 
      (VOC 202012/01) or 20B/501Y.V1, was first identified in the UK 
      and has since been detected in the US 
      and other countries. The growing concern is primarily because 
      it has shown to be significantly more transmissible than other 
      variants. The data presented was gratefully sourced from 
      <a href="https://www.gisaid.org/">GISAID</a>.
    </div>
    <br>
    <div class="alert alert-primary" role="alert">
    As of {{date}}, there's been {{total_num}} sequences
    detected in over {{num_countries}} countries under this strain (see table for definition).
    <br>In the US, there's been {{us_num}} sequences detected in over {{num_states}} states.
    </div>
    <br>
	<!-- </div> -->
      <!-- <div class="row"> -->
        <div class="row align-items-center">
        <div class="col-md-6">
          <table class='table'>
            <caption>Table 1.1: Key Mutations that define the strain</caption>
            <tr>
              <th>Gene</th>
              <th>Nucleotide Mutations</th>
              <th>Amino Acid Changes</th>
            </tr>
            <tr>
              <td>ORF1ab</td>
              <td>C3266T, T6953C, C5387A,
                11288:11296 deletion</td>
              <td>T1001I, I2230T, A1708D</td>
            </tr>
            <tr>
              <td>S</td>
              <td>
                A23062T, C23270A, A23402G, C23603A,
                C23708T, T24505G, G24913C
              <td>DEL69-70, DEL144Y, N501Y, A570D,
              D614G, P681H, T716I, S982A, D1118H</td>
            </tr>
            <tr>
              <td>N</td>
              <td>GAT28279CTA, C28976T</td>
              <td>D3L, S235F</td>
            </tr>
            <tr>
              <td>ORF8</td>
              <td>G28047T, C27971T, A28110G</td>
              <td>R52I, Q27_, Y73C</td>
            </tr>
          </table>
        </div>
        <div class="col-md-6">
          <figure class="figure">
            {{strain_distance_plot}}
            <figcaption class="figure-caption">Figure 1.1: genetic distance (root-to-tip), a measure of evolutionary
               changes, plotted for strain and non-strain (related) samples (excluding other well known VOCs e.g. B.1.135). 
            </figcaption>
          </figure>
        </div>
      </div>
      </div>

      <!-- *** Section 2 - State Prevalence *** --->
      <div class="row" id="section2">
	    <div class="col-6">
          <h3>State Prevalence</h3>
      </div>
      <hr>
        <div class="col-md-6">
          <figure class="figure">
            {{county_map}}
            <figcaption class="figure-caption">Figure 2.1: spatial (geographical) prevalence of the {{strain}} strain across California. 
            </figcaption>
          </figure>
        </div>
        <div class="col-md-6">
          <figure class="figure">
            {{ca_time}}
            <figcaption class="figure-caption">Figure 2.2: temporal (over time) prevalence of the strain across California. 
            </figcaption>
          </figure>
	      </div>
      </div>

      <!-- *** Section 3 - National Prevalence *** --->
      <div class="row" id="section3">
	<div class="col-6">
          <h3>National Prevalence</h3>
	</div>
	<hr>
        <div class="col-md-6">
          <figure class="figure">
            {{state_map}}
            <figcaption class="figure-caption">Figure 3.1: the spatial (geographical) prevalence of the strain across the US.
            </figcaption>
          </figure>
	      </div>
        <div class="col-md-6">
          <figure class="figure">
            {{us_time}}
            <figcaption class="figure-caption">Figure 3.2: the temporal (over time) prevalence of the strain across the US.
            </figcaption>
          </figure>
	      </div>
      </div>

      <!-- *** Section 2 - Global Prevalence *** --->
      <div class="row" id="section4">
	<div class="col-6">
          <h3>Global Prevalence</h3>
          <div class="alert alert-success" role="alert">
              More detailed information on global prevalence of SARS-CoV-2 
              strains of concern can be found in 
              <a href="https://cov-lineages.org/global_report.html">this post</a>.
          </div>
	</div>
  <hr>
        <div class="col-md-6">
          <figure class="figure">
            {{world_map}}
            <figcaption class="figure-caption">Figure: the spatial (geographical) prevalence of the strain 
              across the world.
            </figcaption>
          </figure>
	      </div>
        <div class="col-md-6">
          <figure class="figure">
            {{world_time}}
            <figcaption class="figure-caption">Figure 4.2: the temporal (over time) prevalence of the strain 
              across the world.
            </figcaption>
          </figure>
	      </div>
      </div>


      <!-- *** Section 5 - Notes on Sampling *** --->
      <div class="row" id="section5">
	      <div class="col-6">
          <h3>Notes on Sampling</h3>
	      </div>
	<hr>
        <p>
          As figure 3.2 indicates, the majority of B.1.1.7 genomes identified in the US to-date were identified by S-gene target failures (SGTF) 
          in community-based diagnostic PCR testing. 
          SGTF indicates a deletion mutation that is one of several mutations able to distinguish the B.1.1.7 from other SARS-CoV-2 strains. 
          <br>This tells us the lineage is present in the US. However, since this was not an unbiased sampling approach, 
          it does not indicate the true prevalence of the B117 lineage in the US. 
          Estimates of true prevalence in the US are discussed <a href="https://blog.helix.com/b117-variant-updated-data/">here</a>.
        </p>
  
    <figure class="figure">
      {{ sampling_img }}
      <figcaption class="figure-caption">Figure 5: a simple illustration of how genomic surveillance of 
      COVID-19 samples 
      could allow us to elucidate an increasingly clear picture of how the virus is 
      evolving and spreading. The pictures above are 
      electromagnetic microscopy images of SARS-CoV-2 (credit: <a href="https://www.niaid.nih.gov/news-events/novel-coronavirus-sarscov2-images">NIAID</a>) that are 
      <a href="https://www.biorxiv.org/content/10.1101/740548v8">"crappified"</a> (salt & pepper noise) to varying degrees depending on the rate of COVID-19 sequencing at each location. As a reference, we include a clear picture on the right to indicate that a 5% 
      genomic sampling rate would be an ideal (first) objective to be able to observe 
      statistically significant phenomena.
      across the world.
      </figcaption>
    </figure>
    </div>

      <!-- *** Section 6 - Closing Statements *** --->
      <div class="row" id="section6">
        <div class="col-6">
                <h3>Comments</h3>
        </div>
        <hr>
          <div class="col">
          Research laboratories across the US are encouraged to contribute to COVID-19 genomic sequencing efforts. More detailed information can be found <a href="https://apps.who.int/iris/bitstream/handle/10665/338480/9789240018440-eng.pdf?sequence=1&isAllowed=y">here</a>.
          <br><br>
          Specifically when uploading genomes to GISAID or GenBank, please indicate if the sample was identified via S-gene target failures (SGTF). This can be indicated under the fields "purpose_of_sequencing" (GISAID) or "Additional host information". This will help in identifying the true prevalence of the lineage across the country.
              </div>
              <!-- <div class="col">
                <iframe
                  src="https://giphy.com/embed/U2AJtvSsNCl4oFjWvj" width="480" height="270" frameBorder="0" class="giphy-embed" allowFullScreen>
                </iframe>
                <p><a href="https://giphy.com/gifs/davidmakesman-own-dmm-david-makes-man-U2AJtvSsNCl4oFjWvj">via GIPHY</a></p>
              </div> -->
            </div>
          </div>
  </body>
</html>
